### PR TITLE
Ear 261 new validate piping when deleting

### PR DESCRIPTION
**Pull request by zimboflyman**
### What is the context of this PR?
We need relevant Validation on piping in Question titles
When a previous piped answer is moved or deleted it should show the relevant error message

1. Let authors know if their piped answers have become invalid when they delete an answer which is piped.
2. Let authors know if their piped answers have become invalid when they move a question or section associated with piping.

### How to review 

**1. Deleted Piping**

Create a question/answer A without validation errors
Create question B with piped in Answer A in the title
Delete Question or answer A
You will see a Error message in B - "The answer being piped has been deleted"

**_### Acceptance criteria (user perspective)_**
Delete a question whose answer is piped into another question

GIVEN I'm piping the answer to question A into question B
WHEN I delete the answer to question A (either by deleting the answer itself, or the question)
THEN I get an error message on question B saying "The answer being piped has been deleted"

Delete a question whose answer is piped into another section

GIVEN I'm piping the answer to question A (from section 1) into section 2
WHEN I delete the answer to question A (either by deleting the answer itself, or the question)
THEN I get an error message on section 2 saying "The answer being piped has been deleted"

**2. Moved piping**

Create a question/answer A without validation errors
Create question B with piped in Answer A in the title
Moved Question or answer A AFTER question B
You will see a Error message in B - "The answer being piped is now later in the questionnaire" 

### **_Acceptance criteria (user perspective)_**
Move a question whose answer is piped into another question earlier

GIVEN I'm piping the answer to question A into question B
WHEN I move question B earlier than question A
THEN I get an error message on question B saying "The answer being piped is now later in the questionnaire" 

Move a question which is using a piped answer later 

GIVEN I'm piping the answer to question A into question B
WHEN I move question A later than question B
THEN I get an error message on question B saying "The answer being piped is now later in the questionnaire"

Move a question whose answer is piped into another section later than the section

GIVEN I'm piping the answer to question A (from section 1) into section 2
WHEN I move question A later than section 2
THEN I get an error message on section 2 saying "The answer being piped is now later in the questionnaire"

Move a section which is using a piped answer earlier than the answer being piped 

GIVEN I'm piping the answer to question A into section 2
WHEN I move section 2 earlier than question A
THEN I get an error message on section 2 saying "The answer being piped is now later in the questionnaire"

### What to do after everything is green

1. * [ ] Bring this branch up-to-date with Origin/Master
2. * [ ] If there are a lot of commits or some are not helpful to read, squash them down
3. * [ ] Click the **Merge** button on this pull request
4. * [ ] Does this change mean the **Capability examples** questionnaire should be updated?
5. * [ ] Move the Trello card for this task into the next stage of the process

### CODE REVIEW

File: eq-author-api/src/validation/customKeywords/validatePipingInTitle.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+      dataPath,
+      parentData,
+      fieldName,
+      questionnaire
+    ) {
+      isValid.errors = [];
+
+      const splitDataPath = dataPath.split("/");
+      const currentPage =
+        questionnaire.sections[splitDataPath[2]].pages[splitDataPath[4]];
```
Comment (MattElcock) [inline eq-author-api/src/validation/customKeywords/validatePipingInTitle.js:25]: More of a note to self, really, but we could move these two lines into a util function and update the other custom keyword functions as I'm pretty sure they all (or most of them) use this.

File: eq-author-api/src/validation/customKeywords/validatePipingInTitle.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      const previousAnswersForPage = getPreviousAnswersForPage(
+        questionnaire,
+        currentPage.id,
+        true
+      );
+
+      const $ = cheerio.load(entityData);
+      const pipedIdList = [];
+
+      $("p")
```
Comment (MattElcock) [inline eq-author-api/src/validation/customKeywords/validatePipingInTitle.js:44]: I'm not familiar with Cheerio. Our of curiosity, how come you have "p" in here? Is it to find paragraph tags or something?

Comment (zimboflyman) [reply eq-author-api/src/validation/customKeywords/validatePipingInTitle.js:44]: yes - all the sans that hold each individual piped in Answer or Metdata are contained in a single <p></P>

File: eq-author-api/src/validation/index.test.js [hunk 3]
```diff
@@ -1379,4 +1381,31 @@ describe("schema validation", () => {
...
   });
+
+  describe("Piping validation within Question Labels", () => {
+    it("should validate a piping answer moved after this question", () => {
+      const piping = validation(questionnaire);
+      expect(piping.totalCount).toBe(0);
+
+      questionnaire.sections[0].pages[0].title = `<p><span data-piped="answers" data-id="answer_2" data-type="Number">[number]</span></p>`;
+
+      const Errors = validation(questionnaire);
```
[the remark this answers is not in the document]
Comment (zimboflyman) [reply eq-author-api/src/validation/index.test.js:1392]: Woops !

File: eq-author/src/App/page/Design/CalculatedSummaryPageEditor/index.test.js [hunk 4]
```diff
@@ -4,35 +4,84 @@ import { shallow } from "enzyme";
...
   beforeEach(() => {
-    mockHandlers = {
+    defaultProps = {
       fetchAnswers: jest.fn(),
       onChange: jest.fn(),
       onUpdate: jest.fn(),
       onChangeUpdate: jest.fn(),
       onUpdateCalculatedSummaryPage: jest.fn(),
-      getValidationError: jest.fn(),
+      // getValidationError: jest.fn(),
```
[the remark this answers is not in the document]
Comment (zimboflyman) [reply eq-author/src/App/page/Design/CalculatedSummaryPageEditor/index.test.js:15]: woops!!!

File: eq-author/src/App/page/Design/CalculatedSummaryPageEditor/index.test.js [hunk 5]
```diff
@@ -4,35 +4,84 @@ import { shallow } from "enzyme";
...
+      match: {
+        params: {
+          questionnaireId: "1",
+          sectionId: "2",
+          pageId: "3",
+        },
+      },
+      questionnaireId: "1",
+      // };
```
[the remark this answers is not in the document]
Comment (zimboflyman) [reply eq-author/src/App/page/Design/CalculatedSummaryPageEditor/index.test.js:24]: double woops!

File: eq-author/src/components/Error/ValidationError.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+import PropTypes from "prop-types";
+import styled from "styled-components";
+import { colors } from "constants/theme";
+
+import IconText from "components/IconText";
+import WarningIcon from "constants/icon-warning.svg?inline";
+
+const ErrorStyling = styled(IconText)`
+  color: ${colors.red};
+  width: ${props => (props.right ? "100%" : "80%")};
```
[the remark this answers is not in the document]
Comment (zimboflyman) [reply eq-author/src/components/Error/ValidationError.js:11]: Ahh - this was code bought in from merge master - so not sure i should change it ?
what do we think?
This does bring up issues on larger PR's when merging master into the PR to keep up with master branch....

File: eq-author/src/App/page/Design/CalculatedSummaryPageEditor/index.js [hunk 7]
```diff
@@ -38,6 +38,30 @@ const SelectorTitle = styled.h2`
...
+  },
+  {
+    condition: props =>
+      some(props.page.validationErrorInfo.errors, {
+        errorCode: richTextEditorErrors.PIPING_TITLE_DELETED.errorCode,
+      }),
+    message: () => richTextEditorErrors.PIPING_TITLE_DELETED.message,
+  },
+];
+
```
[the remark this answers is not in the document]
Comment (zimboflyman) [reply eq-author/src/App/page/Design/CalculatedSummaryPageEditor/index.js:64]: Good point - thanks !

File: eq-author-api/src/validation/customKeywords/validatePipingInTitle.js (hunk 1 above)
Comment (zimboflyman) [reply eq-author-api/src/validation/customKeywords/validatePipingInTitle.js:25]: There are 2 similar functions - but one gets id so prob not worth it ?

**Pull request merged**